rfcs/005 - Unsurprising Functions and Scalability RFC.md

```diff
@@ -0,0 +1,215 @@
+
+1. The simplest solution to the RPC behavior is to use [`pgbouncer`](https://www.pgbouncer.org/) to only allow one SQL statement per connection and fail out requests that do otherwise. This is a known configuration that every SQL editor we are aware of can handle, as well as several ORMs. This layer on top of the database should [allow us to enable proper SSL certificates](https://www.percona.com/blog/2021/06/28/enabling-ssl-tls-sessions-in-pgbouncer/) to improve trustworthiness with end users and eliminate (most) MITM possibilities.
+2. Efficiency gains as well as a reduction in complexity can be had by switching to the [`pgsql-http`](https://github.com/pramsey/pgsql-http) extension and having the RPC functions simply wait for the HTTP response from the engine, though they must *all* finish before the HTTP connection times out. This requires getting off of RDS, as this [is not one of the blessed extensions](https://docs.aws.amazon.com/AmazonRDS/latest/PostgreSQLReleaseNotes/postgresql-extensions.html). The HTTP timeout issue can be avoided entirely if the engine is a sidecar process to the Postgres instance and they communicate over `localhost`.
+3. Take the final few HTTP endpoints exposed to the end users and dashboard in the engine and turn most of them into RPC functions in the `metadata` table, and close off HTTP access to the outer world entirely, exposing only the database. The `/run` endpoint would turn into a lambda function for the dashboard, instead. This lets us simplify our deployment artifact into a singular Docker container with all three (Postgres, Engine, pgbouncer) communicating directly to each other.
+4. Deploy this to a horizontally-scaling collection of EC2 instances with EBS volumes for the database, initially only one such instance, but scaling up to more requires a simple routing layer above that decides the target EC2 instance by the database ID being connected to, with the sharding on the number of logical databases per actual database being determined by the scaling factors we run into.
+
+This singular docker container can also be pushed to Dockerhub and make the deployments even simpler, being just downloading a new docker image, restarting the container, and pruning the dead image(s), with deployments taking seconds rather than minutes, though with a brief moment of downtime (before we implement the routing layer, which could hold external connections open until downstream is up again, then making it zero-downtime).
+
+It would also be possible to consider docker-container-per-user-database (with multiple of them per EC2 instance), which would allow us to drop maintaining multiple module versions at the same time within the engine. `iasql_upgrade` would become an *actual* upgrade of the engine to a newer version triggered by the end user, instead, but this and the routing layer could be done progressively in the future after the main transition in the four points above.
+
+### Alternatives Considered
```

Comment:
what about considering a pub/sub approach? https://github.com/pvarentsov/nestjs-pg-notify
or websockets: https://github.com/supabase/realtime

Reply:
So pub/sub is functionally identical to the current approach:

You publish a request and then poll on the response. The only difference I can see is the message passing between the database and engine being two tables instead of one?

The project you linked to is itself built on top of another library called `pg-listen`, the mission-statement of which being that [they didn't have a reliable communication layer](https://github.com/andywer/pg-listen#why-another-package). Graphile Worker *has* been reliable for us in terms of communication between the engine and postgres, so removing it is *not* for that reason. The reasons why we want to remove it are:

1. It adds hundreds of milliseconds of latency to the RPC calls (which are generally not noticeable when AWS API calls take *minutes* but not every cloud API or service is going to be that slow)
2. The current structure causes unexpected behavior from our functions because the requested work is not happening in the connection process.

Switching to a pub/sub approach won't resolve either of those concerns.

Sockets/Websockets *are* an answer that could be considered separate. Since this is our backend database communicating to our backend engine, we could use a standard socket there, and between the dashboard and the engine we could use a websocket.

The socket approach with the database is doable, but we would have to write our own postgres extension instead of re-using an existing one (as SQL doesn't really have a concept of streams, so we'd have to create new extension functions to expose consuming a stream) and then we'd need to write SQL code to parse our streaming protocol out of the socket extension and then write back into it.

It's powerful, but since we just need an RPC where we provide up front all of the data we need to get our response, it feels like overkill.

Also needing a custom Postgres extension kicks us off of RDS, which then makes just using `pgsql_http` on an engine instance running in the same `localhost` the easier and potentially more-reliable approach.

*But* using that project to provide query support to the dashboard looks like it could be much better than the Lambda approach I mentioned. We would need to move the per-statement splitting into the dashboard entirely, but that seems simple enough.

Reply:
I didn't add these into the alternatives considered section because they wouldn't resolve the "next SQL statement on the same connection has access to the new data", at least not on their own, but I incorporated the websocket suggestion into the `/run` replacement text, where I think it is a better fit.